polyply/src/annotate_ligands.py

```diff
@@ -0,0 +1,194 @@
+
+    Returns a dictionary with keys 'mol_name',
+    'mol_idx', and 'resname' for the fields that are specified.
+    Resid will be an int.
+
+    Parameters
+    ----------
+    resspec: str
+    Returns
+    -------
+    dict
+    """
+    molname = None
+    resname = None
+    mol_idx = None
+    resid = None
+    mol_name_idx, *res = resspec.split('-', 1)
+    molname, *mol_idx = mol_name_idx.split('#', 1)
+
+    if res:
+        resname, *resid = res[0].split('#', 1)
+ #   print(mol_idx, resname, molname, resid)
+    out = {}
+    if mol_idx:
+        mol_idx = int(mol_idx[0])
+        out['mol_idx'] = mol_idx
+    if resname:
+        out['resname'] = resname
+    if molname:
+        out['molname'] = molname
+    if resid:
+        out['resid'] = float(resid[0])
+
+    return out
+
+
+def _find_nodes(molecule, mol_attr):
+    nodes = []
+    attr = {}
+    for key in ["resname", "resid"]:
+        if key in mol_attr:
+           attr[key] = mol_attr[key]
+
+    for node in molecule.nodes:
+        if all([molecule.nodes[node][attr] == value for
+                attr, value in attr.items()]):
+
+            nodes.append(node)
+
+    return nodes
+
+class AnnotateLigands(Processor):
+    """
+    Given some molecules defined as Ligands,
+    the processor adds them to the ligated molecules
+    via an edge and indicates the node as ligated.
+    This will cause it to be picked up by the build
+    processor and place the ligands close. The processor
+    also contains an extract method, which will readout
+    the coordinates from the ligated molecules and puts
+    them back such that the topology order is kept.
+    """
+
+    def __init__(self, topology, ligands):
+        """
+        Initalize the processor with a `topology`
+        and a list of `ligands definitions` which
+        are used both in annotating and extracting
+        the positions later.
+        """
+        self.topology = topology
+        self.ligand_defs = defaultdict(list)
+
+        for mol_spec, lig_spec in ligands:
+            mol_attr = parse_residue_spec(mol_spec)
+            lig_attr = parse_residue_spec(lig_spec)
+
+           # check which molecules are elligable for annotation
+            if "mol_idx" in mol_attr:
+                mol_idxs = [mol_attr["mol_idx"]]
+            else:
+                mol_name = mol_attr["molname"]
+                mol_idxs = self.topology.mol_idx_by_name[mol_name]
+
+            # check which ligands are elligable for annotation
+            if "mol_idx" in lig_attr:
+                lig_idxs = [lig_attr["mol_idx"]]
+            else:
+                lig_name = lig_attr["molname"]
+                lig_idxs = self.topology.mol_idx_by_name[lig_name]
+
+            # sanity check
+            if len(lig_idxs) < len(mol_idxs):
+                msg = ("You try to annotate {} molecules with {} ligands "
+                       "However, you can at most assign 1 molecule per ligand. ")
```

Comment:
because that would create a cycle and we do not know how to generate coordinates for cycles

Reply:
Ah ok. 
Note it doesn't *always* create a cycle (could be a bridge between 2 chains), but it may. As for generating cycle coordinates, that may be worth thinking hard over the following days/weeks.

Reply:
We just cannot do cycles. There is no point in thinking hard, I've already explored more aspects than I wanted regarding this matter. Basically what you'd need is a graph embedding algorithm that creates a macro-cycle. Problem is they either:

1) come up with the trivial solution. I.e. they make a circle, which is not what you want generally.
2) are computationally equivalent to running a simulation, at which point there is no advantage of using the program over choosing random coordinates and running a pseudo simulation. 

Then you also run into problems when you have more than 1 cycle, because you need to make sure they don't form knots and need to be packed well enough to allow backmapping (i.e. acceptable ring strain).